compiler/src/iree/compiler/Codegen/Common/GPU/AMDGPULowerCoalescedDMAToGatherLDS.cpp

```diff
@@ -0,0 +1,231 @@
+    int64_t elementBits = sourceType.getElementTypeBitWidth();
+    LDBG() << "  Element type: " << elementType;
+    LDBG() << "  Element bits: " << elementBits;
+
+    int64_t innermostDimSize = sourceType.getShape().back();
+    int64_t transferSizeBits = innermostDimSize * elementBits;
+    LDBG() << "  Source innermost dimension size: " << innermostDimSize;
+    LDBG() << "  Transfer size in bits: " << transferSizeBits;
+
+    std::optional<int64_t> subgroupSize =
+        getSubgroupSize(dmaOp->getParentOfType<FunctionOpInterface>());
+    if (!subgroupSize.has_value()) {
+      return rewriter.notifyMatchFailure(
+          dmaOp, "unable to determine subgroup size from forall");
+    }
+    LDBG() << "Subgroup size: " << *subgroupSize;
+
+    // Check that transfer size matches one of the target DMA sizes.
+    int64_t transferSizePerLane = transferSizeBits / *subgroupSize;
+    LDBG() << "Transfer size per lane: " << transferSizePerLane << " bits";
+
+    if (!targetDmaSizes.empty() &&
+        !llvm::is_contained(targetDmaSizes, transferSizePerLane)) {
+      return rewriter.notifyMatchFailure(
+          dmaOp, "transfer size does not match any target DMA size");
+    }
+    LDBG() << "Transfer size matches target DMA sizes";
+
+    // TODO: Handle indices properly - for now we skip the explicit indices
+    // check The lane parameter is always present but we handle it differently.
+
+    ArrayRef<int64_t> sourceShape = sourceType.getShape();
+    LDBG() << "Source rank: " << sourceShape.size();
+
+    // TODO: Support more general cases such as 1-d.
+    if (sourceShape.size() < 2) {
+      return rewriter.notifyMatchFailure(
+          dmaOp, "source must have at least 2 dimensions");
+    }
+
+    int64_t secondInnermostDimSize = sourceShape[sourceShape.size() - 2];
+    LDBG() << "  Source second innermost dimension size: "
+           << secondInnermostDimSize;
+
+    int64_t elementsPerTransfer = innermostDimSize / *subgroupSize;
+
+    auto transferType = VectorType::get({elementsPerTransfer}, elementType);
+
+    // Actually create the GatherToLDS ops to perform the transfer.
+    rewriter.setInsertionPoint(dmaOp);
+
+    TypedValue<IndexType> laneId = dmaOp.getLane();
+
+    Location loc = dmaOp.getLoc();
+    Value laneOffset = arith::MulIOp::create(
+        rewriter, loc, laneId,
+        arith::ConstantIndexOp::create(rewriter, loc, elementsPerTransfer));
+
+    Value zero = arith::ConstantIndexOp::create(rewriter, loc, 0);
+    for (SmallVector<int64_t> offsets :
+         StaticTileOffsetRange({secondInnermostDimSize}, {1})) {
+      Value iVal = arith::ConstantIndexOp::create(rewriter, loc, offsets[0]);
+
+      SmallVector<Value> srcIndices(sourceType.getRank(), zero);
+      srcIndices[sourceType.getRank() - 2] = iVal;
+      srcIndices[sourceType.getRank() - 1] = laneOffset;
+
+      SmallVector<Value> dstIndices(destType.getRank(), zero);
+      dstIndices[destType.getRank() - 2] = iVal;
```

Comment:
Can you add a few comments explaining the values here and/or what this loop is doing?